### PR TITLE
drivers: usb_dc_rpi_pico: implement remote wakeup support

### PR DESCRIPTION
**Pull request by purdeaandrei**
drivers: usb_dc_rpi_pico: implement remote wakeup support
    
    This change implements the following, necessary for remote wakeup to be
    supported:
    
    - implement usb_dc_wakeup_request() call to trigger remote wakeup
    - implement interrupts to detect supsended/resumed state
    - implement extra logic to simulate resumed state when the resume
      is a result of remote wakeup. In this case the rp2040 chip doesn't
      send a USB_INTR_DEV_RESUME_FROM_HOST interrupt, or any other
      interrupts when the resume condition is detected to be extended
      by the upstream port, so we need to simulate this event when we
      see activity on the bus.

### CODE REVIEW

**Comment by purdeaandrei** [comment]
Note that this PR has https://github.com/zephyrproject-rtos/zephyr/pull/54030 as it's base. For now when reviewing, please look directly at the top-most commit only on this branch. I will rebase, once #54030 is merged.

**Review by tmon-nordic — DISMISSED** [review]

File: drivers/usb/device/usb_dc_rpi_pico.c [hunk 1]
```diff
@@ -236,12 +237,48 @@ static void udc_rpi_handle_buff_status(void)
 	}
 }
 
+static void udc_rpi_handle_resume(void)
+{
+	struct cb_msg msg;
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_rpi_pico.c:242]:
```c
	struct cb_msg msg = {
		.ep = 0U,
		.type = USB_DC_RESUME,
		.ep_event = false,
 	};
```

File: drivers/usb/device/usb_dc_rpi_pico.c [hunk 2]
```diff
@@ -885,6 +935,14 @@ int usb_dc_reset(void)
 	return 0;
 }
 
+int usb_dc_wakeup_request(void)
+{
+	LOG_DBG("Remote Wakeup");
+	state.remote_wakeup_pending = true;
+	hw_set_alias(usb_hw)->sie_ctrl = USB_SIE_CTRL_RESUME_BITS;
+	return 0;
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_rpi_pico.c:943]:
```c
	hw_set_alias(usb_hw)->sie_ctrl = USB_SIE_CTRL_RESUME_BITS;

	return 0;
```

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/usb/device/usb_dc_rpi_pico.c [hunk 3]
```diff
@@ -59,6 +59,7 @@ struct udc_rpi_state {
 	bool should_set_address;
 	uint16_t control_out_ep_rcvd;
 	uint8_t addr;
+	bool remote_wakeup_pending;
```
[the remark this answers is not in the document]
Comment (purdeaandrei) [reply drivers/usb/device/usb_dc_rpi_pico.c]: DONE

File: drivers/usb/device/usb_dc_rpi_pico.c (hunk 1 above)
Comment (purdeaandrei) [reply drivers/usb/device/usb_dc_rpi_pico.c]: DONE

File: drivers/usb/device/usb_dc_rpi_pico.c [hunk 4]
```diff
@@ -236,12 +237,48 @@ static void udc_rpi_handle_buff_status(void)
...
+	msg.type = USB_DC_RESUME;
+	msg.ep_event = false;
+
+	k_msgq_put(&usb_dc_msgq, &msg, K_NO_WAIT);
+	state.remote_wakeup_pending = false;
+}
+
+static void udc_rpi_handle_suspended(void)
+{
+	struct cb_msg msg;
```
[the remark this answers is not in the document]
Comment (purdeaandrei) [reply drivers/usb/device/usb_dc_rpi_pico.c]: DONE

File: drivers/usb/device/usb_dc_rpi_pico.c (hunk 2 above)
Comment (purdeaandrei) [reply drivers/usb/device/usb_dc_rpi_pico.c:947]: DONE

**Comment by purdeaandrei** [comment]
The dependency of this PR is now merged. So I have cleanly rebased this PR on top of main.

**Review by tmon-nordic — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Pull request merged**